File: core/src/main/java/org/evergreen_ils/views/bookbags/BookBagUtils.java
/*
 * Copyright (C) 2016 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

package org.evergreen_ils.views.bookbags;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import org.evergreen_ils.R;
import org.evergreen_ils.accountAccess.AccountAccess;
import org.evergreen_ils.accountAccess.SessionNotFoundException;
import org.evergreen_ils.android.Log;
import org.evergreen_ils.data.BookBag;
import org.evergreen_ils.searchCatalog.RecordInfo;

import java.util.List;

public class BookBagUtils {
    private static final String TAG = BookBagUtils.class.getSimpleName();

    public static void showAddToListDialog(final Activity activity, final List<BookBag> bookBags, final RecordInfo info) {
        final String list_names[] = new String[bookBags.size()];
        for (int i = 0; i < list_names.length; i++)
            list_names[i] = bookBags.get(i).name;

        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setTitle(R.string.choose_list_message);
        builder.setItems(list_names, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                addRecordToList(activity, bookBags.get(which), info);
            }
        });
        builder.create().show();
    }

    public static void addRecordToList(final Activity activity, final BookBag bookbag, final RecordInfo info) {
        final ProgressDialog progressDialog = ProgressDialog.show(activity,
                activity.getString(R.string.dialog_please_wait),
                activity.getString(R.string.adding_to_list_message));
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                AccountAccess ac = AccountAccess.getInstance();
                try {
                    ac.addRecordToBookBag(info.doc_id, bookbag.id);
                } catch (SessionNotFoundException e) {
                    Log.d(TAG, "caught", e);
                }
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        progressDialog.dismiss();
                    }
                });

            }
        });
        thread.start();
    }
}

File: core/src/main/java/org/evergreen_ils/views/bookbags/BookBagUtils.kt
/*
 * Copyright (C) 2016 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */
package org.evergreen_ils.views.bookbags

import android.app.AlertDialog
import android.widget.Toast
import androidx.core.os.bundleOf
import kotlinx.coroutines.async
import net.kenstir.hemlock.R
import net.kenstir.hemlock.android.Analytics
import net.kenstir.hemlock.android.App
import org.evergreen_ils.data.BookBag
import net.kenstir.hemlock.data.Result
import org.evergreen_ils.net.Gateway
import org.evergreen_ils.data.MBRecord
import org.evergreen_ils.utils.ui.BaseActivity
import net.kenstir.hemlock.android.ui.ProgressDialogSupport
import org.evergreen_ils.utils.ui.showAlert

object BookBagUtils {
    fun showAddToListDialog(activity: BaseActivity, bookBags: List<BookBag>, info: MBRecord) {
        val listNames = bookBags.map { it.name }.toTypedArray()

        val builder = AlertDialog.Builder(activity)
        builder.setTitle(R.string.choose_list_message)
        builder.setItems(listNames) { _, which -> addRecordToList(activity, bookBags[which], info) }
        builder.create().show()
    }

    private fun addRecordToList(activity: BaseActivity, bookBag: BookBag, info: MBRecord) {
        activity.scope.async {
            val progress = ProgressDialogSupport()
            try {
                progress.show(activity, activity.getString(R.string.adding_to_list_message))

                val result = Gateway.actor.addItemToBookBagAsync(App.getAccount(), bookBag.id, info.id)
                Analytics.logEvent(Analytics.Event.BOOKBAG_ADD_ITEM, bundleOf(
                    Analytics.Param.RESULT to Analytics.resultValue(result)
                ))
                when (result) {
                    is Result.Success -> {}
                    is Result.Error -> { activity.showAlert(result.exception); return@async }
                }

                Toast.makeText(activity, activity.resources.getString(R.string.msg_added_to_list), Toast.LENGTH_SHORT).show()
            } catch (ex: Exception) {
                activity.showAlert(ex)
            } finally {
                progress.dismiss()
            }
        }
    }
}
